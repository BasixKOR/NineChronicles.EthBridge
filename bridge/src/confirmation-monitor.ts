import { Monitor } from "./monitor";

function delay(ms: number): Promise<void> {
    return new Promise(resolve => {
        setTimeout(() => { resolve() }, ms);
    })
}

export abstract class ConfirmationMonitor<TEventData> extends Monitor<TEventData> {
    private latestBlockNumber: number;

    private readonly _confirmations: number;
    private readonly _delayMilliseconds: number;

    constructor(latestBlockNumber: number, confirmations: number, delayMilliseconds: number = 15 * 1000) {
        super();

        this.latestBlockNumber = latestBlockNumber;

        this._confirmations = confirmations;
        this._delayMilliseconds = delayMilliseconds;
    }

    async * loop(): AsyncIterableIterator<TEventData> {
        while(true) {
            const beforeLatestBlockNumber = this.latestBlockNumber;
            const networkLatestBlockNumber = await this.getTipIndex();
            const confrimedLatestBlockNumber = Math.max(networkLatestBlockNumber - this._confirmations, beforeLatestBlockNumber);

            if (beforeLatestBlockNumber < confrimedLatestBlockNumber) {
                console.debug(`Trying to look up from ${beforeLatestBlockNumber} to ${confrimedLatestBlockNumber}`);
                const eventLogs = await this.getEvents(beforeLatestBlockNumber, confrimedLatestBlockNumber);
                for (const eventLog of eventLogs) {
                    yield eventLog;
                }

                this.latestBlockNumber = confrimedLatestBlockNumber + 1;
            } else {
                console.debug("Skipped...");
            }

            await delay(this._delayMilliseconds);
        }
    }

    protected abstract getTipIndex(): Promise<number>;

    protected abstract getEvents(from: number, to: number): Promise<TEventData[]>;
}
