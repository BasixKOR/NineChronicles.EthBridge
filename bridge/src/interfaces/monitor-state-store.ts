export interface IMonitorStateStore {
    store(network: string, blockHash: string, txId: string): Promise<void>;
    load(network: string): Promise<{ blockHash: string, txId: string }>;
}
