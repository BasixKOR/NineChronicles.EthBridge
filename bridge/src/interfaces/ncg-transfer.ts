import { TxId } from "../types/txid";

export interface INCGTransfer {
    transfer(address: string, amount: BigInt): Promise<TxId>;
}
