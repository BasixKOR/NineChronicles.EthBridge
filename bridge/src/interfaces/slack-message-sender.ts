import { ChatPostMessageArguments, ChatPostMessageResponse } from "@slack/web-api";
import { ForceOmit } from "../types/force-omit";

export interface ISlackMessageSender {
    sendMessage(message: ForceOmit<ChatPostMessageArguments, "channel">): Promise<ChatPostMessageResponse>;
}
