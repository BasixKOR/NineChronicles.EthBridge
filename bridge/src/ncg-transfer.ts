import axios, { AxiosResponse } from "axios";
import { INCGTransfer } from "./interfaces/ncg-transfer";
import { TxId } from "./types/txid";

interface GraphQLRequestBody {
    operationName: string | null;
    query: string;
    variables: Object;
}

export class NCGTransfer implements INCGTransfer {
    private readonly _apiEndpoint: string;

    /**
     * The address of sender;
     */
    private readonly _address: string;

    constructor(apiEndpoint: string, address: string) {
        this._apiEndpoint = apiEndpoint;
        this._address = address;
    }

    async transfer(address: string, amount: BigInt): Promise<TxId> {
        const nextTxNonce = await this.getNextTxNonce(this._address);
        const query = `mutation TransferGold($recipient: Address!, $amount: String!, $txNonce: Long!)
        { transfer(recipient: $recipient, amount: $amount, txNonce: $txNonce) }`;
        const response = await this.graphqlRequest({
            operationName: "TransferGold",
            query,
            variables: {
                recipient: address,
                amount: amount.toString(),
                txNonce: nextTxNonce,
            }});

        return response.data.data.transfer;
    }

    private async getNextTxNonce(address: string): Promise<number> {
        const query = "query GetNextTxNonce($address: Address!) { nextTxNonce(address: $address) } ";
        const response = await this.graphqlRequest({
            operationName: "GetNextTxNonce",
            query,
            variables: { address, }
        });

        return response.data.data.nextTxNonce;
    }

    private async graphqlRequest(body: GraphQLRequestBody): Promise<AxiosResponse> {
        return axios.post(this._apiEndpoint, body,
        {
            headers: {
                "Content-Type": "application/json",
            },
        });
    }
}
