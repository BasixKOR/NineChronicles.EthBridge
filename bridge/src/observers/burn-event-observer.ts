import { IObserver } from ".";
import { BlockHash } from "../types/block-hash";
import { EventData } from "web3-eth-contract";
import { TransactionLocation } from "../types/transaction-location";
import { BurnEventResult } from "../types/burn-event-result";
import { INCGTransfer } from "../interfaces/ncg-transfer";
import { OpenSearchClient } from "../opensearch-client";
import { IMonitorStateStore } from "../interfaces/monitor-state-store";
import { UnwrappedEvent } from "../messages/unwrapped-event";
import Decimal from "decimal.js";
import { UnwrappingFailureEvent } from "../messages/unwrapping-failure-event";
import { Integration } from "../integrations";
import { ISlackMessageSender } from "../interfaces/slack-message-sender";

export class EthereumBurnEventObserver
    implements
        IObserver<{
            blockHash: BlockHash;
            events: (EventData & TransactionLocation)[];
        }>
{
    private readonly _ncgTransfer: INCGTransfer;
    private readonly _opensearchClient: OpenSearchClient;
    private readonly _slackMessageSender: ISlackMessageSender;
    private readonly _monitorStateStore: IMonitorStateStore;
    private readonly _explorerUrl: string;
    private readonly _etherscanUrl: string;
    private readonly _integration: Integration;

    constructor(
        ncgTransfer: INCGTransfer,
        slackMessageSender: ISlackMessageSender,
        opensearchClient: OpenSearchClient,
        monitorStateStore: IMonitorStateStore,
        explorerUrl: string,
        etherscanUrl: string,
        integration: Integration
    ) {
        this._ncgTransfer = ncgTransfer;
        this._slackMessageSender = slackMessageSender;
        this._opensearchClient = opensearchClient;
        this._monitorStateStore = monitorStateStore;
        this._explorerUrl = explorerUrl;
        this._etherscanUrl = etherscanUrl;
        this._integration = integration;
    }

    async notify(data: {
        blockHash: BlockHash;
        events: (EventData & TransactionLocation)[];
    }): Promise<void> {
        const { blockHash, events } = data;
        if (events.length === 0) {
            await this._monitorStateStore.store("ethereum", {
                blockHash,
                txId: null,
            });
        }

        for (const { returnValues, transactionHash, blockHash } of events) {
            const {
                _sender: sender,
                _to,
                amount: burnedWrappedNcgAmountString,
            } = returnValues as BurnEventResult;
            const recipient = _to.substring(0, 42);
            const amount = new Decimal(burnedWrappedNcgAmountString).div(
                new Decimal(10).pow(18)
            );
            const amountString = amount.toFixed(2, Decimal.ROUND_DOWN);

            try {
                console.log("Process Ethereum transaction", transactionHash);
                const nineChroniclesTxId = await this._ncgTransfer.transfer(
                    recipient,
                    amountString,
                    transactionHash
                );

                await this._monitorStateStore.store("ethereum", {
                    blockHash,
                    txId: transactionHash,
                });
                await this._slackMessageSender.sendMessage(
                    new UnwrappedEvent(
                        this._explorerUrl,
                        this._etherscanUrl,
                        sender,
                        recipient,
                        amountString,
                        nineChroniclesTxId,
                        transactionHash
                    ).render()
                );
                await this._opensearchClient.to_opensearch("info", {
                    content: "wNCG -> NCG request success",
                    libplanetTxId: nineChroniclesTxId,
                    ethereumTxId: transactionHash,
                    sender: sender,
                    recipient: recipient,
                    amount: amount.toNumber(),
                });
                console.log("Transferred", nineChroniclesTxId);
            } catch (e) {
                await this._slackMessageSender.sendMessage(
                    new UnwrappingFailureEvent(
                        this._etherscanUrl,
                        sender,
                        recipient,
                        amountString,
                        transactionHash,
                        String(e)
                    ).render()
                );
                await this._opensearchClient.to_opensearch("error", {
                    content: "wNCG -> NCG request failure",
                    cause: String(e),
                    ethereumTxId: transactionHash,
                    sender: sender,
                    recipient: recipient,
                    amount: amount.toNumber(),
                });
                await this._integration.error(
                    "Unexpected error during unwrapping NCG",
                    {
                        errorMessage: String(e),
                        sender,
                        recipient,
                        transactionHash,
                        amountString,
                    }
                );
            }
        }
    }
}
