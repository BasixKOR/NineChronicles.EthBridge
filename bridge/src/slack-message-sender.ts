import {
    ChatPostMessageArguments,
    ChatPostMessageResponse,
    WebClient,
} from "@slack/web-api";
import { ISlackMessageSender } from "./interfaces/slack-message-sender";

export class SlackMessageSender implements ISlackMessageSender {
    private readonly _webClient: WebClient;
    private readonly _channel: string;

    constructor(webClient: WebClient, channel: string) {
        this._webClient = webClient;
        this._channel = channel;
    }

    async sendMessage(
        message: Omit<ChatPostMessageArguments, "channel">
    ): Promise<ChatPostMessageResponse> {
        return this._webClient.chat.postMessage({
            channel: this._channel,
            ...message,
        });
    }
}
