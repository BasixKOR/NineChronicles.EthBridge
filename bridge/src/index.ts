import Web3 from "web3";
import { init } from "@sentry/node";
import { KmsProvider } from "@planetarium/aws-kms-provider";

import { IWrappedNCGMinter } from "./interfaces/wrapped-ncg-minter";
import { INCGTransfer } from "./interfaces/ncg-transfer";
import { EthereumBurnEventMonitor } from "./monitors/ethereum-burn-event-monitor";
import { NineChroniclesTransferredEventMonitor } from "./monitors/nine-chronicles-transferred-event-monitor";
import { WrappedNCGMinter } from "./wrapped-ncg-minter";
import { wNCGTokenAbi } from "./wrapped-ncg-token";
import { HeadlessGraphQLClient } from "./headless-graphql-client";
import { ContractDescription } from "./types/contract-description";
import { IMonitorStateStore } from "./interfaces/monitor-state-store";
import { Sqlite3MonitorStateStore } from "./sqlite3-monitor-state-store";
import { WebClient } from "@slack/web-api"
import { URL } from "url";
import { Configuration } from "./configuration";
import { NCGTransferredEventObserver } from "./observers/nine-chronicles"
import { EthereumBurnEventObserver } from "./observers/burn-event-observer"
import { KMSNCGSigner } from "./kms-ncg-signer";
import { NCGKMSTransfer } from "./ncg-kms-transfer";
import Decimal from "decimal.js";

(async () => {
    const GRAPHQL_API_ENDPOINT: string = Configuration.get("GRAPHQL_API_ENDPOINT");
    const NCG_MINTER: string = Configuration.get("NCG_MINTER");
    const KMS_PROVIDER_URL: string = Configuration.get("KMS_PROVIDER_URL");
    const KMS_PROVIDER_KEY_ID: string = Configuration.get("KMS_PROVIDER_KEY_ID");
    const KMS_PROVIDER_REGION: string = Configuration.get("KMS_PROVIDER_REGION");
    const KMS_PROVIDER_AWS_ACCESSKEY: string = Configuration.get("KMS_PROVIDER_AWS_ACCESSKEY");
    const KMS_PROVIDER_AWS_SECRETKEY: string = Configuration.get("KMS_PROVIDER_AWS_SECRETKEY");
    const KMS_PROVIDER_PUBLIC_KEY: string = Configuration.get("KMS_PROVIDER_PUBLIC_KEY");
    const WNCG_CONTRACT_ADDRESS: string = Configuration.get("WNCG_CONTRACT_ADDRESS");
    const MONITOR_STATE_STORE_PATH: string = Configuration.get("MONITOR_STATE_STORE_PATH");
    const SLACK_WEB_TOKEN: string = Configuration.get("SLACK_WEB_TOKEN");
    const EXPLORER_ROOT_URL: string = Configuration.get("EXPLORER_ROOT_URL");
    const ETHERSCAN_ROOT_URL: string = Configuration.get("ETHERSCAN_ROOT_URL");
    const SENTRY_DSN: string | undefined = Configuration.get("SENTRY_DSN", false);
    if (SENTRY_DSN !== undefined) {
        init({
            dsn: SENTRY_DSN,
        });
    }

    const CONFIRMATIONS = 10;

    const monitorStateStore: IMonitorStateStore = await Sqlite3MonitorStateStore.open(MONITOR_STATE_STORE_PATH);
    const slackWebClient = new WebClient(SLACK_WEB_TOKEN);

    const headlessGraphQLCLient = new HeadlessGraphQLClient(GRAPHQL_API_ENDPOINT);
    const kmsProvider = new KmsProvider(KMS_PROVIDER_URL, {
      region: KMS_PROVIDER_REGION,
      keyIds: [KMS_PROVIDER_KEY_ID],
      credential: {
        accessKeyId: KMS_PROVIDER_AWS_ACCESSKEY,
        secretAccessKey: KMS_PROVIDER_AWS_SECRETKEY
      }
    });
    const web3 = new Web3(kmsProvider);
    const wNCGToken: ContractDescription = {
        abi: wNCGTokenAbi,
        address: WNCG_CONTRACT_ADDRESS,
    };

    if (!web3.utils.isAddress(NCG_MINTER)) {
        throw Error("NCG_MINTER variable seems invalid because it is not valid address format.");
    }

    const kmsAddresses = await kmsProvider.getAccounts();
    if(kmsAddresses.length != 1) {
      throw Error("NineChronicles.EthBridge is supported only one address.");
    }
    const kmsAddress = kmsAddresses[0];
    console.log(kmsAddress);
    const minter: IWrappedNCGMinter = new WrappedNCGMinter(web3, wNCGToken, kmsAddress);
    const signer = new KMSNCGSigner(KMS_PROVIDER_REGION, KMS_PROVIDER_KEY_ID, {
        accessKeyId: KMS_PROVIDER_AWS_ACCESSKEY,
        secretAccessKey: KMS_PROVIDER_AWS_SECRETKEY,
    });
    const derivedAddress = "0x" + web3.utils.keccak256("0x" + Buffer.from(KMS_PROVIDER_PUBLIC_KEY, "base64").toString("hex").slice(2)).slice(26);
    if (kmsAddress.toLowerCase() !== derivedAddress.toLowerCase()) {
        throw Error("KMS_PROVIDER_PUBLIC_KEY variable seems invalid because it doesn't match to address from KMS.");
    }

    const ncgKmsTransfer = new NCGKMSTransfer(
        headlessGraphQLCLient,
        kmsAddress,
        KMS_PROVIDER_PUBLIC_KEY,
        [NCG_MINTER],
        signer
    );

    const ethereumBurnEventObserver = new EthereumBurnEventObserver(ncgKmsTransfer, slackWebClient, monitorStateStore, EXPLORER_ROOT_URL, ETHERSCAN_ROOT_URL);
    const ethereumBurnEventMonitor = new EthereumBurnEventMonitor(web3, wNCGToken, await monitorStateStore.load("ethereum"), CONFIRMATIONS);
    ethereumBurnEventMonitor.attach(ethereumBurnEventObserver);

    const ncgExchangeFeeRatio = new Decimal(0.01);  // 1%
    const ncgTransferredEventObserver = new NCGTransferredEventObserver(ncgKmsTransfer, minter, slackWebClient, monitorStateStore, EXPLORER_ROOT_URL, ETHERSCAN_ROOT_URL, ncgExchangeFeeRatio);
    const nineChroniclesMonitor = new NineChroniclesTransferredEventMonitor(await monitorStateStore.load("nineChronicles"), headlessGraphQLCLient, kmsAddress);
    nineChroniclesMonitor.attach(ncgTransferredEventObserver);

    ethereumBurnEventMonitor.run();
    nineChroniclesMonitor.run();
})().catch(error => {
    console.error(error);
    process.exit(-1);
});
