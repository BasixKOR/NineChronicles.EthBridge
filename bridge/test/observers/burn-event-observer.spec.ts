import { EventData } from "web3-eth-contract";
import { INCGTransfer } from "../../src/interfaces/ncg-transfer";
import { IMonitorStateStore } from "../../src/interfaces/monitor-state-store";
import { OpenSearchClient } from "../../src/opensearch-client";
import { TxId } from "../../src/types/txid";
import { EthereumBurnEventObserver } from "../../src/observers/burn-event-observer";
import { TransactionLocation } from "../../src/types/transaction-location";
import { Integration } from "../../src/integrations";
import { ISlackMessageSender } from "../../src/interfaces/slack-message-sender";
import { SlackMessageSender } from "../../src/slack-message-sender";
import { ISlackChannel } from "../../src/slack-channel";

jest.mock("@slack/web-api", () => {
    return {
        WebClient: jest.fn(() => {
            return {
                chat: {
                    postMessage: jest.fn(),
                },
            };
        }),
    };
});

jest.mock("../../src/opensearch-client", () => {
    return {
        OpenSearchClient: jest.fn(() => {
            return {
                to_opensearch: jest.fn(),
            };
        }),
    };
});

describe(EthereumBurnEventObserver.name, () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    const mockNcgTransfer: jest.Mocked<INCGTransfer> = {
        transfer: jest.fn().mockResolvedValue("TX-HASH"),
    };

    const mockSlackChannel: jest.Mocked<ISlackChannel> = {
        sendMessage: jest.fn(),
    };
    const mockSlackMessageSender: ISlackMessageSender = new SlackMessageSender(
        mockSlackChannel
    );

    const mockOpenSearchClient = new OpenSearchClient(
        "https://www.random-url.com",
        "auth",
        "9c-eth-bridge"
    ) as OpenSearchClient & {
        to_opensearch: ReturnType<typeof jest.fn>;
    };

    const mockMonitorStateStore: jest.Mocked<IMonitorStateStore> = {
        load: jest.fn(),
        store: jest.fn(),
    };

    const mockIntegration: jest.Mocked<Integration> = {
        error: jest.fn(),
    };

    const observer = new EthereumBurnEventObserver(
        mockNcgTransfer,
        mockSlackMessageSender,
        mockOpenSearchClient,
        mockMonitorStateStore,
        "https://explorer.libplanet.io/9c-internal",
        "https://9cscan.com",
        false,
        "https://ropsten.etherscan.io",
        mockIntegration
    );

    describe(EthereumBurnEventObserver.prototype.notify.name, () => {
        it("should record the block hash even if there is no events", () => {
            observer.notify({
                blockHash: "BLOCK-HASH",
                events: [],
            });

            expect(mockMonitorStateStore.store).toHaveBeenCalledWith(
                "ethereum",
                {
                    blockHash: "BLOCK-HASH",
                    txId: null,
                }
            );
        });

        it("should post slack message every events", async () => {
            const ncgRecipient = "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e";
            function makeEvent(
                ncgRecipient: string,
                amount: number,
                txId: TxId
            ): EventData & TransactionLocation {
                return {
                    blockHash: "BLOCK-HASH",
                    address: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                    logIndex: 0,
                    blockNumber: 0,
                    event: "Burn",
                    raw: {
                        data: "",
                        topics: [],
                    },
                    signature: "",
                    transactionIndex: 0,
                    transactionHash: txId,
                    txId: txId,
                    returnValues: {
                        _sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        _to: ncgRecipient,
                        amount: amount,
                    },
                };
            }

            const events = [
                makeEvent(ncgRecipient, 1000000000000000000, "TX-A"),
                makeEvent(ncgRecipient, 1200000000000000000, "TX-B"),
                makeEvent(ncgRecipient, 10000000000000000, "TX-C"),
                makeEvent(ncgRecipient, 3225000000000000000, "TX-D"),
            ];

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events,
            });

            expect(mockMonitorStateStore.store).toHaveBeenCalledWith(
                "ethereum",
                {
                    blockHash: "BLOCK-HASH",
                    txId: "TX-D",
                }
            );

            expect(mockNcgTransfer.transfer.mock.calls).toEqual([
                [ncgRecipient, "1.00", "TX-A"],
                [ncgRecipient, "1.20", "TX-B"],
                [ncgRecipient, "0.01", "TX-C"],
                [ncgRecipient, "3.22", "TX-D"],
            ]);
        });

        it("slack/opensearch message - snapshot", async () => {
            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        blockHash: "BLOCK-HASH",
                        address: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                        logIndex: 0,
                        blockNumber: 0,
                        event: "Burn",
                        raw: {
                            data: "",
                            topics: [],
                        },
                        signature: "",
                        transactionIndex: 0,
                        transactionHash: "TX-ID",
                        txId: "TX-ID",
                        returnValues: {
                            _sender:
                                "0x2734048eC2892d111b4fbAB224400847544FC872",
                            _to: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                            amount: 1000000000000000000,
                        },
                    },
                ],
            });

            expect(
                mockOpenSearchClient.to_opensearch.mock.calls
            ).toMatchSnapshot();
            expect(mockSlackChannel.sendMessage.mock.calls).toMatchSnapshot();
        });

        it("slack/opensearch 9c transfer error message - snapshot", async () => {
            mockNcgTransfer.transfer.mockImplementationOnce(
                (address, amount, memo) => {
                    throw new Error("mockNcgTransfer.transfer error");
                }
            );

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        blockHash: "BLOCK-HASH",
                        address: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                        logIndex: 0,
                        blockNumber: 0,
                        event: "Burn",
                        raw: {
                            data: "",
                            topics: [],
                        },
                        signature: "",
                        transactionIndex: 0,
                        transactionHash: "TX-ID",
                        txId: "TX-ID",
                        returnValues: {
                            _sender:
                                "0x2734048eC2892d111b4fbAB224400847544FC872",
                            _to: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                            amount: 1000000000000000000,
                        },
                    },
                ],
            });

            expect(
                mockOpenSearchClient.to_opensearch.mock.calls
            ).toMatchSnapshot();
            expect(mockSlackChannel.sendMessage.mock.calls).toMatchSnapshot();
        });

        it("pagerduty 9c transfer error message - snapshot", async () => {
            mockNcgTransfer.transfer.mockImplementationOnce(
                (address, amount, memo) => {
                    throw new Error("mockNcgTransfer.transfer error");
                }
            );

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        blockHash: "BLOCK-HASH",
                        address: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                        logIndex: 0,
                        blockNumber: 0,
                        event: "Burn",
                        raw: {
                            data: "",
                            topics: [],
                        },
                        signature: "",
                        transactionIndex: 0,
                        transactionHash: "TX-ID",
                        txId: "TX-ID",
                        returnValues: {
                            _sender:
                                "0x2734048eC2892d111b4fbAB224400847544FC872",
                            _to: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                            amount: 1000000000000000000,
                        },
                    },
                ],
            });

            expect(mockIntegration.error.mock.calls).toMatchSnapshot();
        });
    });
});
