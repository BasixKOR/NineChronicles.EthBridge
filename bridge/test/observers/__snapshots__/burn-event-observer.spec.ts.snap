// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthereumBurnEventObserver notify slack message - snapshot 1`] = `
Array [
  Array [
    Object {
      "attachments": Array [
        Object {
          "author_name": "Bridge Event",
          "color": "#42f5aa",
          "fallback": "wNCG 0x2734048eC2892d111b4fbAB224400847544FC872 → NCG 0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
          "fields": Array [
            Object {
              "title": "9c network transaction",
              "value": "https://explorer.libplanet.io/9c-internal/transaction/?TX-HASH",
            },
            Object {
              "title": "Ethereum network transaction",
              "value": "https://ropsten.etherscan.io/tx/TX-ID?",
            },
            Object {
              "title": "sender (Ethereum)",
              "value": "0x2734048eC2892d111b4fbAB224400847544FC872",
            },
            Object {
              "title": "recipient (NineChronicles)",
              "value": "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
            },
            Object {
              "title": "amount",
              "value": "1.00",
            },
          ],
        },
      ],
      "channel": "#nine-chronicles-bridge-bot",
      "text": "wNCG → NCG event occurred.",
    },
  ],
]
`;
